### PR TITLE
Fail loudly with corrupted Parquet statistics

### PR DESCRIPTION
**Pull request by zhenxiao**
@nezihyigitbasi @dain @findepi @qqibrow

### CODE REVIEW

**Comment by zhenxiao** [comment]
thank you, @electrum 
I get comments addressed. will do following task to add dataSourceID

**Comment by nezihyigitbasi** [comment]
In addition to knowing which file is bad, it would be easier to debug if we know which column's stats is corrupted. By looking at the stack trace of the exception we can see for which type we have corrupted stats, but if we have many columns of that type then that's not useful. In short, an exception message like: `Corrupted statistics for column %s in Parquet file %s` would be pretty useful.

**Comment by nezihyigitbasi** [comment]
There is also a risk of breaking user queries that scan Parquet files with corrupted stats. Do we want to have a flag to revert to the current behavior just in case (which we can eventually remove and make this new behavior the default behavior)?

**Comment by findepi** [comment]
> Do we want to have a flag to revert to the current behavior just in case (which we can eventually remove and make this new behavior the default behavior)?

@nezihyigitbasi IMO we should fail on corrupted stats by default -- unless some widely used Parquet writing program is notorious for writing corrupted stats. 
And yes, we should have safety flag to disable the check, at least for some transition period.

**Comment by zhenxiao** [comment]
@nezihyigitbasi @findepi @electrum thanks ur comments
get comments addressed, added dataSourceId for Parquet, added configuration to control whether fail query when scan Parquet file with corrupted statistics. During transition period, disable the check, we could enable it, and probably remove the config, to fail whenever hitting corrupted statistics

**Comment by nezihyigitbasi** [comment]
> IMO we should fail on corrupted stats by default -- unless some widely used Parquet writing program is notorious for writing corrupted stats.

@findepi My comment was confusing probably, sorry. I totally agree with you. We should fail on corrupt stats by default. @zhenxiao looks like this PR sets the flag to false by default btw.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 1]
```diff
@@ -903,6 +904,19 @@ public HiveClientConfig setUseParquetColumnNames(boolean useParquetColumnNames)
         return this;
     }
 
+    public boolean isFailOnCorruptedParquetStatistics()
+    {
+        return failOnCorruptedParquetStatistics;
+    }
+
+    @Config("hive.parquet.fail-on-corrupted-statistics")
+    @ConfigDescription("Fail query when scan Parquet file with corrupted statistics")
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: `Fail when scanning Parquet files with corrupted statistics`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java [hunk 2]
```diff
@@ -225,6 +226,11 @@ public HiveSessionProperties(HiveClientConfig hiveClientConfig, OrcFileWriterCon
                         "Experimental: Parquet: Access Parquet columns using names from the file",
                         hiveClientConfig.isUseParquetColumnNames(),
                         false),
+                booleanProperty(
+                        PARQUET_FAIL_WITH_CORRUPTED_STATISTICS,
+                        "Parquet: Fail query when scan Parquet file with corrupted statistics",
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java]: "Fail when scanning Parquet files with corrupted statistics"

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/HdfsParquetDataSource.java [hunk 3]
```diff
@@ -89,15 +97,15 @@ private void readInternal(long position, byte[] buffer, int bufferOffset, int bu
             throw e;
         }
         catch (Exception e) {
-            throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error reading from %s at position %s", name, position), e);
+            throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error reading from %s at position %s", id.toString(), position), e);
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/HdfsParquetDataSource.java]: Don't need the `toString` call.

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/Predicate.java [hunk 4]
```diff
@@ -41,8 +42,10 @@ public boolean matches(Map<ColumnDescriptor, DictionaryDescriptor> dictionaries)
      * @param numberOfRows the number of rows in the segment; this can be used with
      * Statistics to determine if a column is only null
      * @param statistics column statistics
+     * @param id Parquet file name
+     * @param failOnCorruptedParquetStatistics whether to fail query when scan Parquet file with corrupted statistics
```
Comment (nezihyigitbasi) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/Predicate.java]: `when scanning a Parquet file`

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 5]
```diff
@@ -149,16 +151,20 @@ else if ((type.equals(BIGINT) || type.equals(TINYINT) || type.equals(SMALLINT) |
...
+                    }
                     return Domain.create(ValueSet.all(type), hasNullValue);
                 }
                 parquetIntegerStatistics = new ParquetIntegerStatistics(longStatistics.genericGetMin(), longStatistics.genericGetMax());
             }
             else {
                 IntStatistics intStatistics = (IntStatistics) statistics;
-                // ignore corrupted statistics
                 if (intStatistics.genericGetMin() > intStatistics.genericGetMax()) {
+                    if (failOnCorruptedParquetStatistics) {
```
Comment (nezihyigitbasi) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java]: You can create a method like this one (feel free to find a better name) to simplify & get rid of repetition:

```
private static void failWithCorruptionException(boolean failOnCorruptedParquetStatistics, String column, ParquetDataSourceId id, Object min, Object max)
{
    if (failOnCorruptedParquetStatistics) {
        throw new ParquetCorruptionException(format("Corrupted statistics for column %s in Parquet file %s: min %s, max %s", column, id, min, max));
    }
}
```

Then calls will look like:

```
failWithCorruptionException(failOnCorruptedParquetStatistics, column, id, intStatistics.genericGetMin(), intStatistics.genericGetMax());
```

**Comment by zhenxiao** [comment]
thank you @nezihyigitbasi 
I get comments addressed, squash into one commit, added testcases for ignore and fail on corrupted statistics

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 6]
```diff
@@ -280,6 +282,13 @@ else if (isVarcharType(type) && columnDescriptor.getType() == PrimitiveTypeName.
         return null;
     }
 
+    private static void failWithCorruptionException(boolean failOnCorruptedParquetStatistics, String column, ParquetDataSourceId id, Statistics statistics)
+    {
+        if (failOnCorruptedParquetStatistics) {
+            throw new ParquetCorruptionException(format("Corrupted statistics for column %s in Parquet file %s: %s", column, id, statistics));
```
Comment (nezihyigitbasi) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java]: I hope all children of `Statistics` type have `toString` implemented.

Comment (nezihyigitbasi) [reply presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java]: One minor change we can make to the error message is to add quotes/brackets to set the boundaries of the column/file name and stats information:

```
Corrupted statistics for column "column_name" in Parquet file "file_name": [stats_string]
```

Currently it looks like the following, which is a little bit hard to follow:

```
Corrupted statistics for column DateColumn in Parquet file testFile: min: 200, max: 100, num_nulls: 0
```

File: presto-parquet/src/main/java/com/facebook/presto/parquet/ParquetCorruptionException.java [hunk 7]
```diff
@@ -13,12 +13,10 @@
  */
 package com.facebook.presto.parquet;
 
-import java.io.IOException;
-
 import static java.lang.String.format;
 
 public class ParquetCorruptionException
-        extends IOException
```
[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-parquet/src/main/java/com/facebook/presto/parquet/ParquetCorruptionException.java]: OK, will keep it

**Comment by zhenxiao** [comment]
thank you, @nezihyigitbasi 
comments addressed, added testcases for all types

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 8]
```diff
@@ -205,6 +211,9 @@ public static ParquetPageSource createParquetPageSource(
             if (e instanceof PrestoException) {
                 throw (PrestoException) e;
             }
+            if (nullToEmpty(e.getMessage()).trim().startsWith("Corrupted statistics for column")) {
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: Instead of relying on the error message, can't we just do:

```
if (e instanceof ParquetCorruptionException)
    throw new PrestoException(HIVE_BAD_DATA, e);
}
```

**Comment by zhenxiao** [comment]
thank you, @nezihyigitbasi 
get comments addressed
@dain @electrum @findepi your comments are welcome

**Comment by nezihyigitbasi** [comment]
Thanks @zhenxiao, merged.
